### DISCUSSION TITLE
Detecting query subsets

### DISCUSSION THREAD

**Discussion by robertlagrant** [sqlalchemy/sqlalchemy#6547] [Usage Questions]
Hello!

I am creating a dynamic query builder and wondered if it's possible to detect if one query is a subset of another.

E.g.
```python
session.query(Users).filter(users.first_name.in_(["rob", "greg"]))
```

is a subset of
```python
session.query(Users).filter(users.first_name.in_(["rob"]))
```

Is this possible using SQLAlchemy's ORM or Core?

**Comment by zzzeek** [comment]
my impression is that you want to do this all on the Python side, which is doable for simpler cases as a graph problem, though I would advise ditching session.query() and using straight select() constructs all the way as they are traversible, which is what you'd need to solve this problem.  The Query() object can turn itself into a select() from the .statement accessor, that might be good enough here.

If you just wanted to answer this question about two queries by any means necessary, you can get a 100% accurate answer by combining the two queries into an EXCEPT statement on the database side; EXCEPT is in the same set operation category as the more familiar UNION / UNION ALL except that it will tell you what rows are different.

solving the issue as a Python-side graph problem can work for the example you have above, but I'm not sure what kinds of complications arise when the WHERE conditions are more complicated / nested / etc.    Things like additoinal tables / JOINs at this level are essentially represented as further WHERE conditions.  Also if one query or the other used aggregates / HAVING / etc. it becomes a weird problem again, so I think the biggest Q with this problem is what are the up front assertions about the problem space, are we given two SELECT statements that have the identical COLUMNS / FROM clause in all cases up front?   

So I think a generalized solution to this would be hard, but just to give you familiarity with the APIs that you can use to walk through this, below is an example of the very specific situation you have above.   **this requires SQLAlchemy 1.4**, as the ".compare()" method was not really implemented prior to 1.4.

```python
import itertools

from sqlalchemy import Column
from sqlalchemy import Integer
from sqlalchemy import select
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.sql import operators
from sqlalchemy.sql import visitors
from sqlalchemy.sql.expression import BinaryExpression

Base = declarative_base()


class A(Base):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)
    data = Column(String)


stmt1 = select(A).where(A.data.in_(["rob", "greg"]))
stmt2 = select(A).where(A.data.in_(["rob"]))

for obj1, obj2 in itertools.zip_longest(
    visitors.iterate(stmt1),
    visitors.iterate(stmt2),
):
    if (
        isinstance(obj1, BinaryExpression)
        and obj1.operator is operators.in_op
        and not obj1.compare(obj2)
    ):
        print(
            "comparing %s and %s"
            % (
                obj1.compile(compile_kwargs={"literal_binds": True}),
                obj2.compile(compile_kwargs={"literal_binds": True}),
            )
        )
        if set(obj2.right.value).issubset(obj1.right.value):
            print("it's a subset")
        else:
            print("it's not a subset")
```

**Reply by robertlagrant** [reply]
Thank you Mike; this is extremely useful and interesting. My feeling is this would work because our filtering system can likely be translated into relatively predictable query code. I will play and report back.
